src/libraries/System.Net.Requests/src/System/Net/HttpWebRequest.cs

```diff
@@ -1125,98 +1134,123 @@ public override Stream EndGetRequestStream(IAsyncResult asyncResult)
-                }
+            if (_servicePoint?.Expect100Continue == true)
+            {
+                _sendRequestMessage.Headers.ExpectContinue = true;
+            }
+
+            _sendRequestMessage.Version = ProtocolVersion;
 
-                request.Version = ProtocolVersion;
+            _sendRequestMessage.Headers.TransferEncodingChunked = SendChunked;
+
+            if (KeepAlive)
+            {
+                _sendRequestMessage.Headers.Connection.Add(HttpKnownHeaderNames.KeepAlive);
+            }
+            else
+            {
+                _sendRequestMessage.Headers.ConnectionClose = true;
+            }
 
-                _sendRequestTask = async ?
-                    client.SendAsync(request, _allowReadStreamBuffering ? HttpCompletionOption.ResponseContentRead : HttpCompletionOption.ResponseHeadersRead, _sendRequestCts!.Token) :
-                    Task.FromResult(client.Send(request, _allowReadStreamBuffering ? HttpCompletionOption.ResponseContentRead : HttpCompletionOption.ResponseHeadersRead, _sendRequestCts!.Token));
+            _sendRequestMessage.Version = ProtocolVersion;
+            HttpCompletionOption completionOption = _allowReadStreamBuffering ?
+                            HttpCompletionOption.ResponseContentRead : HttpCompletionOption.ResponseHeadersRead;
+            _sendRequestTask = async ?
+                _httpClient.SendAsync(_sendRequestMessage, completionOption, _sendRequestCts!.Token) :
+                Task.Run(() => _httpClient.Send(_sendRequestMessage, completionOption, _sendRequestCts!.Token));
 
-                HttpResponseMessage responseMessage = await _sendRequestTask.ConfigureAwait(false);
+            return _sendRequestTask!;
+        }
 
-                HttpWebResponse response = new HttpWebResponse(responseMessage, _requestUri, _cookieContainer);
+        private async Task<WebResponse> HandleResponse(bool async)
+        {
+            _sendRequestTask ??= SendRequest(async);
 
-                int maxSuccessStatusCode = AllowAutoRedirect ? 299 : 399;
-                if ((int)response.StatusCode > maxSuccessStatusCode || (int)response.StatusCode < 200)
+            try
+            {
+                if (_requestStream is not null) // If we didn't flush and end the request stream, do it now, don't wait for dispose.
                 {
-                    throw new WebException(
-                        SR.Format(SR.net_servererror, (int)response.StatusCode, response.StatusDescription),
-                        null,
-                        WebExceptionStatus.ProtocolError,
-                        response);
+                    if (async)
+                    {
+                        await _requestStream!.FlushAsync().ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        _requestStream!.Flush();
+                    }
+                    _requestStream!.EndWriteOnStreamBuffer();
                 }
```

Comment:
is it a given that the user will be done writing by this point?

That is, would something like this not have worked before?
```c#
var request = ...;
request.AllowWriteStreamBuffering = true;

using var requestStrem = await request.GetRequestStreamAsync();
Task responseTask = request.GetResponseAsync();

await requestStream.WriteAsync(foo);
await requestStream.FlushAsync();

await responseTask;
```

Reply:
I believe once we call, `GetResponse` we're basically saying that `hey there is no more data I want to send`, so it should be safe to call it here.